### DISCUSSION TITLE
Return `Vec<Option<Type>` with `belonging_to`

### DISCUSSION THREAD

**Discussion by RubberDuckShobe** [diesel-rs/diesel#4402] [Q&A] [answered]
Hello, I have an *optional* 1-to-1 relation set up between `Song` and `ExtraSongInfo`.
When I want to get a single `Song`'s `ExtraSongInfo`, I can do this and it will return an `Option<ExtraSongInfo>`:

```rust
let extra_info: Option<ExtraSongInfo> = ExtraSongInfo::belonging_to(&song)
    .first(&mut conn)
    .await
    .optional()?;
```

Now, I want to do this for a `Vec<Song>`, where songs may not have an `ExtraSongInfo` associated with them.
So, I tried this:

```rust
let extra_info = ExtraSongInfo::belonging_to(&songs)
    .select(ExtraSongInfo::as_select())
    .load(&mut conn)
    .await
    .optional()?;
```

This, however, returns `Option<Vec<ExtraSongInfo>>`.
Side note: When I omit `.optional()`, Visual Studio Code (or rust-analyzer itself?) tries telling me that `extra_info` is `Vec<Option<ExtraSongInfo>>`, even though it's actually `Vec<ExtraSongInfo>`.

How can I actually get `Vec<Option<ExtraSongInfo>>` where it's `None` for every song that doesn't have an associated `ExtraSongInfo` row?

**Comment by weiznich** [comment, marked as answer]
The correct way to model an 1:1 relation is using a join. For optional 1:1 dependencies that means you want to use `left_join()`. 

In this particular case you end up with a query like that:

```rust
songs::table.left_join(extra_song_info::table).filter(songs::id.eq_any(song_ids)).select(Option::<ExtraSongInfo>::as_select()).load(&mut conn)?;
```

See the [Relation Guide](https://diesel.rs/guides/relations.html) for more details and information on this topic.

> Side note: When I omit .optional(), Visual Studio Code (or rust-analyzer itself?) tries telling me that extra_info is Vec<Option<ExtraSongInfo>>, even though it's actually Vec<ExtraSongInfo>.

That's an issue in rust-analyzer, not in diesel. Please report it in their issue tracker, as we cannot do much about it.

**Comment by RubberDuckShobe** [comment]
That works, thank you!
